### PR TITLE
Convert proxy to go, update Makefiles and Dockerfiles, and update test

### PR DESCRIPTION
**Pull request by ikehz**
Ref kubernetes/kubernetes#8867.

### CODE REVIEW

File: metadata_proxy.go [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+
+var (
+	concealedEndpoints = []string{
+		"/0.1/meta-data/attributes/kube-env",
+		"/computeMetadata/v1beta1/instance/attributes/kube-env",
+		"/computeMetadata/v1/instance/attributes/kube-env",
+	}
+	concealedPatterns = []*regexp.Regexp{
+		regexp.MustCompile("/0.1/meta-data/service-accounts/.+/identity"),
+		regexp.MustCompile("/computeMetadata/v1beta1/instance/service-accounts/.+/identity"),
```
Comment (cjcullen) [inline metadata_proxy.go:20]: Should we block out other versions (e.g. "v2alpha1") with these regexps?

File: metadata_proxy.go [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+		"/0.1/meta-data",
+		"/computeMetadata",
+		"/computeMetadata/",
+		"/computeMetadata/v1beta1",
+		"/computeMetadata/v1",
+	}
+)
+
+func main() {
+	log.Fatal(http.ListenAndServe("127.0.0.1:988", newMetadataHandler()))
```
Comment (cjcullen) [inline metadata_proxy.go]: Should the port be configurable?

File: metadata_proxy.go [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+func newMetadataHandler() http.Handler {
+	u, err := url.Parse("http://169.254.169.254")
+	if err != nil {
+		log.Fatal(err)
+	}
+	proxy := httputil.NewSingleHostReverseProxy(u)
+
+	x := xForwardedForStripper{}
+	proxy.Transport = x
```
Comment (cjcullen) [inline metadata_proxy.go]: Don't really need the intermediate variable "x"

File: metadata_proxy.go [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+	}
+}
+
+func (h *metadataHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	log.Println(req.URL.Path)
+	if req.URL.Query().Get("recursive") != "" {
+		http.Error(rw, "?recursive calls are not allowed by the metadata proxy.", http.StatusForbidden)
+		return
+	}
+	for _, e := range concealedEndpoints {
```
Comment (cjcullen) [inline metadata_proxy.go]: I think the order of these checks is intentional based on what we expect to hit most often. Can you make that clear so it doesn't get lost if this code gets jostled.

File: metadata_proxy.go (hunk 1 above)
Comment (ikehz) [reply metadata_proxy.go:20]: Addressed in comments below.

File: metadata_proxy.go (hunk 2 above)
Comment (ikehz) [reply metadata_proxy.go]: Added TODO.

File: metadata_proxy.go (hunk 3 above)
Comment (ikehz) [reply metadata_proxy.go]: Done.

File: metadata_proxy.go (hunk 4 above)
Comment (ikehz) [reply metadata_proxy.go]: Done.

**Comment by ikehz** [comment]
@cjcullen ptal.

File: metadata_proxy.go [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+		http.Error(rw, "Calls with X-Forwarded-For header are not allowed by the metadata proxy.", http.StatusForbidden)
+	}
+	// Check that the request isn't a recursive one.
+	if req.URL.Query().Get("recursive") != "" {
+		http.Error(rw, "?recursive calls are not allowed by the metadata proxy.", http.StatusForbidden)
+		return
+	}
+
+	// Conceal kube-env and vm identity endpoints for known API versions.
+	// Don't block unkown API versions, since we don't know if have the
```
Comment (cjcullen) [inline metadata_proxy.go]: unknown

Comment (cjcullen) [inline metadata_proxy.go]: "if they have the..."

**Pull request merged**